url7f729337e3_timmolter_Sundial_tgz-pJ8-org_knowm_sundial_SampleRunJ8/src/org/quartz/core/RAMJobStore.java

```diff
@@ -0,0 +1,844 @@
+   * </p>
+   */
+  @Override
+  public List<TriggerFiredResult> triggersFired(List<OperableTrigger> triggers) {
+
+    synchronized (lock) {
+      List<TriggerFiredResult> results = new ArrayList<TriggerFiredResult>();
+
+      for (OperableTrigger trigger : triggers) {
+        TriggerWrapper tw = wrappedTriggersByKey.get(trigger.getName());
+        // was the trigger deleted since being acquired?
+        if (tw == null || tw.trigger == null) {
+          continue;
+        }
+        // was the trigger completed, paused, blocked, etc. since being acquired?
+        if (tw.state != TriggerWrapper.STATE_ACQUIRED) {
+          continue;
+        }
+
+        Calendar cal = null;
+        if (tw.trigger.getCalendarName() != null) {
+          cal = retrieveCalendar(tw.trigger.getCalendarName());
+          if (cal == null) {
+            continue;
+          }
+        }
+        Date prevFireTime = trigger.getPreviousFireTime();
+        // in case trigger was replaced between acquiring and firing
+        timeWrappedTriggers.remove(tw);
+        // call triggered on our copy, and the scheduler's copy
+        tw.trigger.triggered(cal);
+        trigger.triggered(cal);
+        // tw.state = TriggerWrapper.STATE_EXECUTING;
+        tw.state = TriggerWrapper.STATE_WAITING;
+
+        TriggerFiredBundle bndle = new TriggerFiredBundle(retrieveJob(tw.jobKey), trigger, cal, false, new Date(), trigger.getPreviousFireTime(),
+            prevFireTime, trigger.getNextFireTime());
+
+        JobDetail job = bndle.getJobDetail();
+
+        if (!job.isConcurrencyAllowed()) {
+          ArrayList<TriggerWrapper> trigs = getTriggerWrappersForJob(job.getName());
+          Iterator<TriggerWrapper> itr = trigs.iterator();
+          while (itr.hasNext()) {
+            TriggerWrapper ttw = itr.next();
+            if (ttw.state == TriggerWrapper.STATE_WAITING) {
+              ttw.state = TriggerWrapper.STATE_BLOCKED;
+            }
+            if (ttw.state == TriggerWrapper.STATE_PAUSED) {
+              ttw.state = TriggerWrapper.STATE_PAUSED_BLOCKED;
+            }
+            timeWrappedTriggers.remove(ttw);
+          }
+          blockedJobs.add(job.getName());
+        } else if (tw.trigger.getNextFireTime() != null) {
+          synchronized (lock) {
+            timeWrappedTriggers.add(tw);
+          }
+        }
+
+        results.add(new TriggerFiredResult(bndle));
+      }
+      return results;
+    }
+  }
+
+  /**
+   * <p>
+   * Inform the <code>JobStore</code> that the scheduler has completed the firing of the given <code>Trigger</code> (and the execution its associated
+   * <code>Job</code>), and that the <code>{@link org.quartz.jobs.JobDataMap}</code> in the given <code>JobDetail</code> should be updated if the
+   * <code>Job</code> is stateful.
+   * </p>
+   */
+  @Override
+  public void triggeredJobComplete(OperableTrigger trigger, JobDetail jobDetail, CompletedExecutionInstruction triggerInstCode) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.